### PR TITLE
Add LevyFlightSampler based on Lévy-flight random walks

### PR DESCRIPTION
**Pull request by AdityaChopra18**
## Contributor Agreements
Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.
- [x] I agree to the contributor agreements.

> [!TIP]
> Please follow the [Quick TODO list](https://github.com/optuna/optunahub-registry/tree/main?tab=readme-ov-file#quick-todo-list-towards-contribution) to smoothly merge your PR.

## Motivation

Standard samplers like TPE perform well in many settings but can struggle to escape local optima in highly multi-modal search spaces. Lévy flights — characterized by a heavy-tailed step-length distribution — are a well-established mechanism in metaheuristic research for balancing local exploitation with occasional long-range exploration. This PR brings that capability to Optuna users as a drop-in sampler backed by a mathematically sound implementation.

## Description of the changes

This PR adds a new package under `samplers/levy_flight_sampler/`:

- **Algorithm:** Implements the Mantegna algorithm ([Mantegna, 1994](https://doi.org/10.1103/PhysRevE.49.4677)) to generate Lévy stable random steps efficiently using standard Gaussian variates, avoiding the need for heavy numerical integration.
- **Reference:** Based on Yang & Deb (2010), *"Engineering Optimisation by Cuckoo Search"* ([DOI](https://doi.org/10.1504/IJMMNO.2010.035430)), which formalizes the use of Lévy flights for black-box optimization.
- **Distribution support:** Handles `FloatDistribution`, `IntDistribution`, and `CategoricalDistribution`, with native support for log-scale parameters on numerical distributions.
- **Cold-start logic:** Automatically falls back to random sampling for the first two trials before Lévy steps can be computed.
- **Testing:** Ships with a comprehensive test suite of 26 passing tests covering boundary conditions, statistical properties, and all supported distribution types.
- **Example:** Includes `example.py` benchmarking the sampler against TPE on a shifted Rosenbrock function.

## TODO List towards PR Merge

- [x] Copy `./template/` to create your package
- [x] Replace `<COPYRIGHT HOLDER>` in `LICENSE` of your package with your name
- [x] Fill out `README.md` in your package
- [x] Add import statements of your function or class names to be used in `__init__.py`
- [x] Add `from __future__ import annotations` at the head of any Python files that include typing to support older Python versions
- [x] Apply the formatter based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)
- [x] Check whether your module works as intended based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main) 

## References

- Yang, X.-S. & Deb, S. (2010). *Engineering Optimisation by Cuckoo Search*.
  International Journal of Mathematical Modelling and Numerical Optimisation, 1(4), 330–343.
  https://doi.org/10.1504/IJMMNO.2010.035430

- Mantegna, R. N. (1994). *Fast, accurate algorithm for numerical simulation of Lévy stable stochastic processes*.
  Physical Review E, 49(5), 4677.
  https://doi.org/10.1103/PhysRevE.49.4677

### CODE REVIEW

**Comment by not522** [comment]
@y0z Could you review this PR?

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                    new_val = float(np.clip(new_val, low, high))
+
+                if dist.step is not None:
+                    # Round to the nearest grid point.
+                    n_steps = round((new_val - low) / dist.step)
+                    new_val = float(np.clip(low + n_steps * dist.step, low, high))
+
+                params[name] = new_val
+
+            elif isinstance(dist, IntDistribution):
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py:161]: `suggest_int` also has the `log` option ([see](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial.suggest_int)). It would be helpful to handle this appropriately or inform the user that it is not supported.

File: package/samplers/levy_flight_sampler/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+| Parameter    | Default | Effect                                                                                 |
+| ------------ | ------- | -------------------------------------------------------------------------------------- |
+| `beta`       | `1.5`   | Tail heaviness of the Lévy distribution.  Range `(0, 2]`; `2` ≈ Gaussian, `1` ≈ Cauchy |
+| `step_scale` | `0.1`   | Step size as a fraction of each parameter's range                                      |
+| `seed`       | `None`  | RNG seed for reproducibility                                                           |
+
+## Installation
+
+```bash
+pip install optunahub
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/README.md]:
```suggestion
pip install optunahub numpy
```

It is also helpful to include a `requirements.txt` file. Then, you can also write as follows.
```
pip install -r https://hub.optuna.org/samplers/levy_flight_sampler/requirements.txt
```

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+        # Not enough history — fall back to independent (random) sampling.
+        if len(completed) < 2:
+            return {}
+
+        # Find the current best trial.
+        if study.direction == optuna.study.StudyDirection.MINIMIZE:
+            best_trial = min(completed, key=lambda t: t.value)  # type: ignore[arg-type]
+        else:
+            best_trial = max(completed, key=lambda t: t.value)  # type: ignore[arg-type]
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py]: Why don't you use `best_trial`? It is useful since it handles the study direction internally. https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.best_trial

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    ) -> dict[str, BaseDistribution]:
+        return optuna.search_space.intersection_search_space(study.get_trials(deepcopy=False))
+
+    def sample_relative(
+        self,
+        study: optuna.Study,
+        trial: optuna.trial.FrozenTrial,
+        search_space: dict[str, BaseDistribution],
+    ) -> dict[str, Any]:
+        if not search_space:
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py:116]: Based on my understanding, this sampler is for single-objective optimization. Then, you can notify users of this using [`_raise_error_of_multi_objective`](https://github.com/optuna/optuna/blob/master/optuna/samplers/_base.py#L231).

```suggestion
        self._raise_error_if_multi_objective(study)

        if not search_space:
```

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+        return params
+
+    def sample_independent(
+        self,
+        study: optuna.Study,
+        trial: optuna.trial.FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py:199]: Ditto.
```suggestion
    ) -> Any:
    self._raise_error_if_multi_objective(study)
    
```

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    def _levy_step(self, search_range: float) -> float:
+        """Draw a single Lévy-flight step scaled to ``search_range``.
+
+        Uses the Mantegna algorithm to approximate a Lévy stable variate
+        efficiently without requiring special-function libraries.
+        """
+        # Mantegna algorithm: u ~ N(0, sigma^2), v ~ N(0, 1)
+        u = self._rng.normal(0.0, self._sigma)
+        v = self._rng.normal(0.0, 1.0)
+        step = u / (abs(v) ** (1.0 / self._beta))
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py]: Question: Is it possible for `v` to become 0, resulting in a zero div?

**Review by y0z — COMMENTED** [review]
Thank you for the interesting sampler implementation. I was able to test it and didn't encounter any major problems. I'll leave some minor review comments for further refinement.

File: package/samplers/levy_flight_sampler/README.md (hunk 2 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/README.md]: Added requirements.txt and updated README install instructions.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 3 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py]: Switched to `study.best_trial` — cleaner and handles direction internally.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 6 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py]: Added a guard — if v==0, replace with `np.finfo(float).tiny` before division.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 5 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py:199]: Added `self._raise_error_if_multi_objective(study)` to both `sample_relative` and `sample_independent`.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 4 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py:116]: Added `self._raise_error_if_multi_objective(study)` to both `sample_relative` and `sample_independent`.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 1 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py:161]: Added log-space handling for IntDistribution, same approach as FloatDistribution.

**Comment by AdityaChopra18** [comment]
Thanks for the feedback! All comments addressed in the latest commits — added `_raise_error_if_multi_objective,` switched to `study.best_trial,` handled `IntDistribution` log, guarded against `v=0,` and added `requirements.txt`.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    @staticmethod
+    def _mantegna_sigma(beta: float) -> float:
+        """Compute the Mantegna sigma parameter for the given stability index.
+
+        sigma = ( Gamma(1+beta)*sin(pi*beta/2) /
+                  (Gamma((1+beta)/2) * beta * 2^((beta-1)/2)) )^(1/beta)
+
+        This is numerically stable for beta in (0, 2].
+        """
+        num = math.gamma(1.0 + beta) * math.sin(math.pi * beta / 2.0)
```
Comment (y0z) [inline package/samplers/levy_flight_sampler/levy_flight_sampler.py]: Although I'm not very familiar with Mantegna’s (1994) Lévy formulation, I have a concern about the behavior in the boundary case, i.e., `beta = 2.0` (in the original paper, called `alpha`), then `math.sin(math.pi * beta / 2.0)` becomes `sin(π) = 0`. Below is a quick answer from ChatGPT. Does the current implementation work properly?

Image

**Comment by y0z** [comment]
Thank you very much for your updates!
> Thanks for the feedback! All comments addressed in the latest commits — added _raise_error_if_multi_objective, switched to study.best_trial, handled IntDistribution log, guarded against v=0, and added requirements.txt.

File: package/samplers/levy_flight_sampler/levy_flight_sampler.py (hunk 7 above)
Comment (AdityaChopra18) [reply package/samplers/levy_flight_sampler/levy_flight_sampler.py]: You're right to flag this. I investigated and confirmed the bug — at `beta=2.0,` Mantegna's sigma collapses to ~0, meaning all steps were essentially zero and the sampler got stuck.
Fixed in the latest commit: when `beta=2.0,` `_levy_step` now samples directly from `N(0, 1)` instead of using the Mantegna formula — which is the mathematically correct behaviour at the Gaussian limit. The empirical step distribution now has kurtosis ≈ 0, confirming it's properly Gaussian.
Also updated `_mantegna_sigma` to only be called for `beta` < 2.0.

**Review by y0z — APPROVED** [review]
LGTM. Thank you for your contribution!

**Comment by y0z** [comment]
Published:
https://hub.optuna.org/samplers/levy_flight_sampler/

**Comment by AdityaChopra18** [comment]
Thank you for the review and the merge! Really enjoyed the process — learned a lot from your feedback, especially the beta=2.0 boundary case.
